[Logs UI] Respect and report on log threshold alert limits

weltenwort opened the pull request:
## :memo: Summary

This changes the log threshold rule executor such that it respects the alert limit and reports if it has been reached.

closes #141925

## :female_detective: Review notes

- The cancellation mentioned in the issue's ACs is already implemented in the ES client itself (see #124374).
- We're not applying the limits to the query, because we only intend to limit the number of alerts and not the number of checked groups. The query structure currently doesn't allow for expressing that limitation in the query.

matschaffer commented on x-pack/plugins/infra/server/lib/alerting/log_threshold/log_threshold_executor.test.ts:
```diff
@@ -436,12 +443,19 @@ describe('Log threshold executor', () => {
           },
         ]);
+
+        expect(alertLimitMock.setLimitReached).toHaveBeenCalledWith(false);
```
The naming of the spec doesn't make it clear to me why ungrouped would be `setLimitReached: false` and grouped would be `setLimitReached: true`. But I may be lacking some common context.

weltenwort replied:
Good point, I'll look into splitting up the test cases.

matschaffer commented on x-pack/test/api_integration/apis/logs_ui/log_threshold_alert.ts:
```diff
@@ -125,6 +138,7 @@ export default function ({ getService }: FtrProviderContext) {
               },
             ],
           ]);
+          expect(alertLimit.setLimitReached.calledOnceWith(true)).to.be(true);
```
Similar situation here. `with group by should work` doesn't tell me anything about the expected limit behavior.

matschaffer reviewed (approved):
I was wondering if we should maybe have limit-dedicated tests to make it clear why a given test was reaching the limit (or not).

But the code looks sound and meets the AC, so going ahead with approval.

I didn't do any manual testing, but I'd be happy to try whatever testing steps you recommend.

matschaffer commented:
@elasticmachine merge upstream

matschaffer commented on x-pack/plugins/infra/server/lib/alerting/log_threshold/log_threshold_executor.test.ts:
```diff
@@ -514,6 +601,110 @@ describe('Log threshold executor', () => {
         ]);
       });
+
+      it('respects and reports reaching a low limit when alerting', () => {
```
I'm liking these new tests. Makes easier to see what's going in and compare it to the asserted results.

weltenwort commented:
@matschaffer Thanks for the reviews so far! I didn't want to replicate the exact same checks in the functional tests that the unit test already performs. So instead I renamed the existing ones for clarity and added one that specifically tests the case when the limit is below the number of alerts that would be fired without the limit.

Does that work for you?

matschaffer reviewed (approved):
Re-upping approval. The only thing that stood out for me was the duplication of the `Two groups should fire, one shouldn't` test scenario structures. But given it's only two tests, I'd say it can stay and we'll worry about extraction if we ever end up adding a 3rd.
